File: packages/cortex-dart/test/resources/chat_test.dart
import 'dart:convert';

import 'package:nfinitmonkeys_cortex_sdk/cortex_sdk.dart';
import 'package:http/http.dart' as http;
import 'package:test/test.dart';

import '../helpers.dart';

void main() {
  group('ChatCompletionsResource', () {
    late CortexClient client;

    tearDown(() => client.close());

    test('create returns ChatCompletion', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, sampleChatCompletion()),
      );

      final response = await client.chat.completions.create(
        model: 'gpt-4',
        messages: [const ChatMessage.user('Hello')],
      );

      expect(response.id, 'chatcmpl-123');
      expect(response.model, 'gpt-4');
      expect(response.choices.first.message.content,
          'Hello! How can I help you?');
      expect(response.usage!.totalTokens, 21);
    });

    test('create sends correct request body', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(
            200, sampleChatCompletion(), requests),
      );

      await client.chat.completions.create(
        model: 'gpt-4',
        messages: [
          const ChatMessage.system('You are helpful'),
          const ChatMessage.user('Hello'),
        ],
        temperature: 0.7,
        maxTokens: 100,
      );

      expect(requests, hasLength(1));
      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['model'], 'gpt-4');
      expect(body['temperature'], 0.7);
      expect(body['max_tokens'], 100);
      expect(body['stream'], false);
      expect((body['messages'] as List).length, 2);
    });

    test('create sends Authorization header', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(
            200, sampleChatCompletion(), requests),
      );

      await client.chat.completions.create(
        model: 'gpt-4',
        messages: [const ChatMessage.user('Hello')],
      );

      expect(
        requests.first.headers['Authorization'],
        'Bearer sk-cortex-test-key-1234',
      );
    });

    test('create validates empty model', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.chat.completions.create(
          model: '',
          messages: [const ChatMessage.user('Hello')],
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('create validates empty messages', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.chat.completions.create(
          model: 'gpt-4',
          messages: [],
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('create validates temperature range', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.chat.completions.create(
          model: 'gpt-4',
          messages: [const ChatMessage.user('Hello')],
          temperature: 3.0,
        ),
        throwsA(isA<CortexValidationException>()),
      );

      expect(
        () => client.chat.completions.create(
          model: 'gpt-4',
          messages: [const ChatMessage.user('Hello')],
          temperature: -1.0,
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('create validates topP range', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.chat.completions.create(
          model: 'gpt-4',
          messages: [const ChatMessage.user('Hello')],
          topP: 1.5,
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('create validates presencePenalty range', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.chat.completions.create(
          model: 'gpt-4',
          messages: [const ChatMessage.user('Hello')],
          presencePenalty: 3.0,
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('create validates frequencyPenalty range', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.chat.completions.create(
          model: 'gpt-4',
          messages: [const ChatMessage.user('Hello')],
          frequencyPenalty: -3.0,
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('create with tools', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(
            200, sampleChatCompletion(), requests),
      );

      await client.chat.completions.create(
        model: 'gpt-4',
        messages: [const ChatMessage.user('What is the weather?')],
        tools: [
          ToolDefinition(
            function_: const FunctionDefinition(
              name: 'get_weather',
              description: 'Get weather for a city',
              parameters: {
                'type': 'object',
                'properties': {
                  'city': {'type': 'string'},
                },
              },
            ),
          ),
        ],
      );

      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['tools'], hasLength(1));
    });

    test('create handles 401 error', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        maxRetries: 0,
        httpClient: mockClient(401, {'error': 'Invalid API key'}),
      );

      expect(
        () => client.chat.completions.create(
          model: 'gpt-4',
          messages: [const ChatMessage.user('Hello')],
        ),
        throwsA(isA<CortexAuthenticationException>()),
      );
    });

    test('create handles 429 error', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        maxRetries: 0,
        httpClient: mockClient(429, 'Rate limited'),
      );

      expect(
        () => client.chat.completions.create(
          model: 'gpt-4',
          messages: [const ChatMessage.user('Hello')],
        ),
        throwsA(isA<CortexRateLimitException>()),
      );
    });

    test('createStream returns Stream of chunks', () async {
      final sseBody = sampleSseStream(
        contentChunks: ['Hello', ' world', '!'],
      );

      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: StreamingMockClient(sseBody),
      );

      final chunks = await client.chat.completions
          .createStream(
            model: 'gpt-4',
            messages: [const ChatMessage.user('Hello')],
          )
          .toList();

      expect(chunks, hasLength(3));
      expect(chunks[0].choices.first.delta?.content, 'Hello');
      expect(chunks[1].choices.first.delta?.content, ' world');
      expect(chunks[2].choices.first.delta?.content, '!');
    });

    test('createStream validates parameters', () {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.chat.completions.createStream(
          model: '',
          messages: [const ChatMessage.user('Hello')],
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });
  });
}

File: packages/cortex-js/tests/resources/chat.test.ts
import { describe, it, expect } from 'vitest';
import { ValidationError } from '../../src/errors.js';
import { mockFetch, mockStreamFetch, createClient } from '../helpers.js';

const MOCK_CHAT_RESPONSE = {
  id: 'chatcmpl-abc123',
  object: 'chat.completion',
  created: 1700000000,
  model: 'gpt-4',
  choices: [
    {
      index: 0,
      message: { role: 'assistant', content: 'Hello! How can I help you?' },
      finish_reason: 'stop',
    },
  ],
  usage: { prompt_tokens: 10, completion_tokens: 8, total_tokens: 18 },
};

describe('chat.completions', () => {
  describe('create (non-streaming)', () => {
    it('sends a chat completion request', async () => {
      const fetch = mockFetch(MOCK_CHAT_RESPONSE);
      const client = createClient(fetch);

      const result = await client.chat.completions.create({
        model: 'gpt-4',
        messages: [{ role: 'user', content: 'Hello' }],
      });

      expect(result.id).toBe('chatcmpl-abc123');
      expect(result.choices[0]!.message.content).toBe('Hello! How can I help you?');
      expect(result.usage!.total_tokens).toBe(18);

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://api.test.com/v1/chat/completions');
      expect(opts.method).toBe('POST');
      expect(JSON.parse(opts.body as string)).toEqual({
        model: 'gpt-4',
        messages: [{ role: 'user', content: 'Hello' }],
      });
    });

    it('sends all optional parameters', async () => {
      const fetch = mockFetch(MOCK_CHAT_RESPONSE);
      const client = createClient(fetch);

      await client.chat.completions.create({
        model: 'gpt-4',
        messages: [{ role: 'user', content: 'Hello' }],
        temperature: 0.7,
        max_tokens: 100,
        top_p: 0.9,
        stop: ['\n'],
        presence_penalty: 0.5,
        frequency_penalty: 0.5,
        user: 'user-123',
      });

      const body = JSON.parse((fetch.mock.calls[0] as [string, RequestInit])[1].body as string);
      expect(body.temperature).toBe(0.7);
      expect(body.max_tokens).toBe(100);
      expect(body.top_p).toBe(0.9);
      expect(body.stop).toEqual(['\n']);
    });

    it('throws ValidationError when model is missing', async () => {
      const client = createClient(mockFetch({}));

      await expect(
        client.chat.completions.create({
          model: '',
          messages: [{ role: 'user', content: 'Hello' }],
        }),
      ).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when messages is empty', async () => {
      const client = createClient(mockFetch({}));

      await expect(
        client.chat.completions.create({
          model: 'gpt-4',
          messages: [],
        }),
      ).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when message has no role', async () => {
      const client = createClient(mockFetch({}));

      await expect(
        client.chat.completions.create({
          model: 'gpt-4',
          messages: [{ role: '' as 'user', content: 'Hello' }],
        }),
      ).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when temperature is out of range', async () => {
      const client = createClient(mockFetch({}));

      await expect(
        client.chat.completions.create({
          model: 'gpt-4',
          messages: [{ role: 'user', content: 'Hello' }],
          temperature: 3,
        }),
      ).rejects.toThrow(ValidationError);

      await expect(
        client.chat.completions.create({
          model: 'gpt-4',
          messages: [{ role: 'user', content: 'Hello' }],
          temperature: -1,
        }),
      ).rejects.toThrow(ValidationError);
    });
  });

  describe('create (streaming)', () => {
    it('returns an SSE stream', async () => {
      const chunk = JSON.stringify({
        id: 'chatcmpl-1',
        object: 'chat.completion.chunk',
        created: 1700000000,
        model: 'gpt-4',
        choices: [{ index: 0, delta: { content: 'Hello' }, finish_reason: null }],
      });

      const fetch = mockStreamFetch([`data: ${chunk}\n\ndata: [DONE]\n\n`]);
      const client = createClient(fetch);

      const stream = await client.chat.completions.create({
        model: 'gpt-4',
        messages: [{ role: 'user', content: 'Hi' }],
        stream: true,
      });

      const chunks = [];
      for await (const c of stream) {
        chunks.push(c);
      }

      expect(chunks).toHaveLength(1);
      expect(chunks[0]!.choices[0]!.delta.content).toBe('Hello');

      // Verify Accept header was set for streaming
      const [, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect((opts.headers as Record<string, string>)['Accept']).toBe('text/event-stream');
    });
  });
});
